unstructured_ingest/v2/processes/connectors/gitlab.py

```diff
@@ -0,0 +1,267 @@
+        project = gitlab.projects.get(self.repo_path)
+
+        logger.info(f"Successfully accessed project '{self.repo_path}'")
+        return project
+
+
+class GitLabIndexerConfig(IndexerConfig):
+    path: Path = Field(
+        default="/", description=("Path to the location in the repository that will be processed.")
+    )
+    recursive: bool = Field(
+        default=True,
+        description=(
+            "Flag to control recursive operations when indexing. "
+            "If True, the indexer will traverse directories recursively."
+        ),
+    )
+    git_branch: Optional[str] = Field(
+        default=None,
+        description="The name of the branch to interact with.",
+    )
+
+
+@dataclass
+class GitLabIndexer(Indexer):
+    connection_config: GitLabConnectionConfig
+    index_config: GitLabIndexerConfig
+
+    def precheck(self) -> None:
+        """Validates the connection to the GitLab instance by authenticating or
+        accessing the project.
+
+        This method ensures that the GitLab credentials and configuration are correct by
+        either authenticating or attempting to fetch the specified project.
+
+        Raises:
+            SourceConnectionError: If the connection or authentication with GitLab fails.
+        """
+
+        try:
+            gitlab = self.connection_config.get_client()
+            if self.connection_config.access_config.get_secret_value().access_token is not None:
+                gitlab.auth()
+            else:
+                gitlab.projects.get(self.connection_config.repo_path)
+
+        except Exception as e:
+            logger.error(f"Failed to validate connection: {e}", exc_info=True)
+            raise SourceConnectionError(f"Failed to validate connection: {e}")
+
+    def run(self, **kwargs: Any) -> Generator[FileData, None, None]:
+        """Iterates over the GitLab repository tree and yields file metadata as `FileData` objects.
+
+        This method fetches the repository tree for the specified branch and iterates
+        over its contents. For each file (blob), it generates a `FileData` object containing
+        the file's metadata, path, and permissions.
+
+        Args:
+            **kwargs (Any): Additional keyword arguments (if required).
+
+        Yields:
+            FileData: A generator that yields `FileData` objects representing each file (blob)
+            in the repository.
+        """
+        project = self.connection_config.get_project()
+
+        ref = self.index_config.git_branch or project.default_branch
+
+        files = project.repository_tree(
+            path=str(self.index_config.path),
+            ref=ref,
+            recursive=self.index_config.recursive,
+            iterator=True,
+            all=True,
+        )
+
+        for file in files:
+            relative_path = str(Path(file["path"]).relative_to(self.index_config.path))
+            if file["type"] == "blob":
```

Comment:
Curious, what else could the `type` value be here? A directory?

Reply:
It can be "tree" for a directory and I believe "commit" for a submodule but I've not verified the second one yet.